src/structs.py

```diff
@@ -1093,6 +1092,49 @@ def __post_init__(self) -> None:
         assert len(self.states) == len(self.responses) == len(self.actions) + 1
 
 
+@dataclass(frozen=True, eq=False, repr=False)
+class Query(abc.ABC):
+    """Base class for a Query.
+
+    Has no API.
+    """
+
+
+@dataclass(frozen=True, eq=False, repr=False)
+class Response(abc.ABC):
+    """Base class for a Response to a query.
+
+    All responses contain the Query object itself, for convenience.
+    """
+    query: Query
+
+
+@dataclass(frozen=True, eq=False, repr=False)
+class GroundAtomsHoldQuery(Query):
+    """A query for whether ground atoms hold in the state."""
+    ground_atoms: Collection[GroundAtom]
+
+
+@dataclass(frozen=True, eq=False, repr=False)
+class GroundAtomsHoldResponse(Response):
+    """A response to a GroundAtomsHoldQuery, providing boolean answers."""
+    holds: Dict[GroundAtom, bool]
+
+
+@dataclass(frozen=True, eq=False, repr=False)
+class DemonstrationQuery(Query):
+    """A query requesting a demonstration to get from a certain state to a
```

Comment:
"from a certain state to a goal" -> "from the state to a certain goal"
remember the query is about a goal, the agent doesn't get to ask about any state